core/src/global/ionic-global.ts

```diff
@@ -1,16 +1,114 @@
+    `Invalid mode and theme combination provided: mode: ${mode}, theme: ${theme}. Fallback mode ${getDefaultModeForTheme(
+      theme
+    )} will be used.`,
+    ref
+  );
+};
+
+/**
+ * Validates if a mode is accepted for a theme configuration.
+ * @param mode The mode to validate.
+ * @param theme The theme the mode is being used with.
+ * @returns `true` if the mode is valid for the theme, `false` if invalid.
+ */
+const isModeValidForTheme = (mode: Mode, theme: Theme) => {
+  if (mode === 'md') {
+    return theme === 'md' || theme === 'ionic';
+  } else if (mode === 'ios') {
+    return theme === 'ios' || theme === 'ionic';
+  }
+  return false;
+};
 
-export const getIonMode = (ref?: any): Mode => {
-  return (ref && getMode(ref)) || defaultMode;
+/**
+ * Returns the default mode for a specified theme.
+ * @param theme The theme to return a default mode for.
+ * @returns The default mode, either `ios` or `md`.
+ */
+const getDefaultModeForTheme = (theme: Theme): Mode => {
+  if (theme === 'ios') {
+    return 'ios';
+  }
+  return 'md';
+};
+
+/**
+ * Returns the default theme for a specified mode.
+ * @param mode The mode to return a default theme for.
+ * @returns The default theme.
+ */
+const getDefaultThemeForMode = (mode: Mode): Theme => {
+  if (mode === 'ios') {
+    return 'ios';
+  }
+  return 'md';
+};
+
+const isModeSupported = (elmMode: string) => ['ios', 'md'].includes(elmMode);
+
+const isThemeSupported = (theme: string) => ['ios', 'md', 'ionic'].includes(theme);
+
+const isIonicElement = (elm: HTMLElement) => elm.tagName?.startsWith('ION-');
+
+/**
+ * Returns the mode value of the element reference or the closest
+ * parent with a valid mode.
+ * @param ref The element reference to look up the mode for.
+ * @param theme Optionally can provide the theme to avoid an additional look-up.
+ * @returns The mode value for the element reference.
+ */
+export const getIonMode = (ref?: any, theme = getIonTheme(ref)): Mode => {
+  if (ref?.mode && isModeValidForTheme(ref?.mode, theme)) {
+    /**
+     * If the reference already has a mode configuration,
+     * use it instead of performing a look-up.
+     */
+    return ref.mode;
+  } else {
+    const el = getElement(ref);
+    const mode = (el.closest('[mode]')?.getAttribute('mode') as Mode) || defaultMode;
+
+    if (isModeValidForTheme(mode, theme)) {
+      /**
+       * The mode configuration is supported for the configured theme.
+       */
+      return mode;
+    } else {
+      printInvalidModeWarning(mode, theme, ref);
+    }
+  }
+  return getDefaultModeForTheme(theme);
+};
+
+/**
+ * Returns the theme value of the element reference or the closest
+ * parent with a valid theme.
+ *
+ * @param ref The element reference to look up the theme for.
+ * @returns The theme value for the element reference, defaults to
+ * the default theme if it cannot be determined.
+ */
+export const getIonTheme = (ref?: any): Theme => {
+  return (ref && getMode(ref)) || defaultTheme;
```

Comment:
The design doc specified that if the `theme` isn't given, the fallback should match the `mode`. Right now, it's always defaulting to `md`.